### PR TITLE
Calendar External Object support

### PR DESCRIPTION
**Pull request by mcrispino**

### CODE REVIEW

File: src/gxcore/sd/calendar/schedule.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    "T" +
+    stringFromTimeNumber(dt.getUTCHours()) +
+    stringFromTimeNumber(dt.getUTCMinutes()) +
+    stringFromTimeNumber(dt.getUTCSeconds()) +
+    "Z"
+  );
+}
+
+function stringFromTimeNumber(num: number): string {
+  return (num < 10 ? "0" : "") + num.toString();
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  return num.toString().padStart(2, '0');
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  return result;
+}
+
+export function createCalendarEvent(
+  start: Date,
+  end: Date,
+  title: string,
+  place: string
+): string {
+  return (
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]: I think you could take advantage of [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) here

File: src/gxcore/sd/calendar/schedule.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+ */
+export const schedule = (
+  title: string,
+  startDate: Date,
+  endDate: Date,
+  startTime: Date,
+  endTime: Date,
+  place: string
+) => {
+  let sDateTime = getDateTime(startDate, startTime);
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const sDateTime = getDateTime(startDate, startTime);
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+  let eDateTime = getDateTime(endDate, endTime, addHours(sDateTime, 1));
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const eDateTime = getDateTime(endDate, endTime, addHours(sDateTime, 1));
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+  let calStr = createCalendarEvent(sDateTime, eDateTime, title, place);
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const calStr = createCalendarEvent(sDateTime, eDateTime, title, place);
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+function getDateTime(
+  datePart: Date,
+  timePart: Date,
+  defaultValue?: Date
+): Date {
+  if (!datePart) {
+    return defaultValue;
+  }
+
+  let result = datePart;
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const result = datePart;
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    "END:VCALENDAR"
+  );
+}
+
+function getGUIDString(): string {
+  return GUID.newGuid().toString();
+}
+
+function downloadCalendar(calendarStr: string) {
+  let guidStr = getGUIDString();
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const guidStr = getGUIDString();
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+  let fileName = guidStr + ".ics";
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const fileName = `${guidStr}.ics`;
```

File: src/gxcore/sd/calendar/schedule.ts [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+  var element = document.createElement("a");
```
Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]: Setting `window.location.href` didn't work?

Comment (ncardeli) [inline src/gxcore/sd/calendar/schedule.ts]:
```suggestion
  const element = document.createElement("a");
```

Comment (mcrispino) [reply src/gxcore/sd/calendar/schedule.ts]: It didn't, does not do anything...

**Pull request merged**